### ISSUE TITLE
Antd message doesn't work in remote app

### ISSUE THREAD

**Issue by wkei** [originjs/vite-plugin-federation#267]
### Versions

- vite-plugin-federation: 1.1.6
- vite: 2.9.13

### Reproduction

https://stackblitz.com/edit/vitejs-vite-r6clfz?file=apps%2Fcontainer%2Fsrc%2FApp.tsx,README.md&terminal=dev

### Steps to reproduce

Click `message` button, and check console for error message.

PS: However, if we also use `antd/message` or `antd/notification` in the host app, the remote app will also work.

### What is Expected?

`antd/message` popup successfully

### What is actually happening?

`antd/message` or `antd/notification` don't work in vite federation remote app

it throw `reactRender is not a function` error

**Comment by wkei** [comment]
similar issue https://github.com/originjs/vite-plugin-federation/issues/90

**Comment by zgid123** [comment]
Hi @wkei, I am looking at your issue

so the expectation result will be like this?

<img width="1253" alt="image" src="https://user-images.githubusercontent.com/15771072/199712097-79d6db52-ebec-44d7-be17-e66184ebd1ca.png">

**Comment by zgid123** [comment]
I don't have idea why it has this issue. I am looking at the `message` of `antd`. It may be a problem of `antd`, not this package issue

**Comment by zgid123** [comment]
the main problem of your issue is from this one

```js
import { importShared } from './__federation_fn_import.js';
const {
  default: ReactDOM,
  r: reactDom,
  R: ReactDOM$1,
} = await importShared('react-dom');

const { default: React, r: react } = await importShared('react');
```

there are three instance imported from `react-dom`, and the last one `R: ReactDOM$1` for some reasons is undefined when remote app (the app installs the antd), but if we run that app alone, that instance is not undefined.

**Comment by zgid123** [comment]
```js
var ReactDOM = reactDom.exports;

var ReactDOM$1 = /*#__PURE__*/_mergeNamespaces({
  __proto__: null,
  'default': ReactDOM
}, [reactDom.exports]);

export { ReactDOM$1 as R, ReactDOM as default, reactDom as r };
```

inside `__federation_shared_react-dom.js` in your remote app (with-antd) has those code line, I still don't know why we have `ReactDOM$1` here. This one will be undefined if we run remote app inside host app.

I noted here, I will continue debugging later when I have time. Currently my project has some issues, and I cannot spend much time to help

**Comment by zgid123** [comment]
@wkei I think your issue is related with my issue. You can check the status of the issue in #278

**Comment by elpddev** [comment]
This also happens with `@emotion/react` and `@mantine/core`. 

Two react apps, the shared use `emotion`, the host does not. The two react federation bundles are different in the `ReactDOM$1`. Only the shared has it. The host does not, making the shared fail when loaded inside the host.

**Comment by zgid123** [comment]
ok, I will conclude this issue. I had this conclusion two weeks ago, but I forgot to post it.

So, this issue occurs is because the remote app use more usages from the package than the host app.

I don't know how to explain the thing above. But I will try to explain it by all the things I already investigated.

For my case, I use `chakra-ui`, it is using `@emotion/react`. For my host app, when I built it, the generated file only contains `useDisclosure` - check #278. But remote app, it will contain `useDisclosure` and `useDisclosure$1`. It is because I import `useDisclosure` for a drawer and also use a tooltip which using `useDisclosure` too. Then this package builds our code, it will export those two and import those ones for the `index<hash>.js` file.

But my host app only contains `useDisclosure`, that why when the module federation downloads the remote files, and uses it for the main app, we don't have `useDisclosure$1` (because we share it). For the author case, it is `ReactDOM` from a package of `antd`.

For my case, to work around, I will create an other `App` and wrap the content with a `ChakraProvider`, it will act as different theme provider with the host app, and I don't share the `chakra-ui` package for host app and remote app.

Now whenever module federation downloads the files from remote app, it will use those files as package's source instead of the files from host app. It won't raise error anymore.
For my side project, it is acceptable because my remote app may have a different theme with my host app. So the context is correct. For your app, I don't know.

For my opinion, we cannot fix this issue at the moment (I need more investigates from webpack). Because we need to force all the packages use the same dependencies (like our case is `ReactDOM` or/and `useDisclosure`), and we need to make sure the export name must be like the export name from host app's files.
I don't know if webpack has this issue or not. If not, how webpack resolves those problems. If someone already config module federation of webpack for the author's source code, plz share it to me, so I can continue my investigation.

To summarize, to fix this issue, we can have those options:

1. force the built files to use the same objects from packages. For my case, we will force it to use `useDisclosure` instead of import `useDisclosure` and `useDisclosure$1`
2. we keep the current progress and if the host app import `useDisclosure$1` and the host app's files does not have that object, we will force it to use remote app's files instead

* solution 1: we will have pit fall. They are:
- what if those two objects come from the same package but are imported in different version, how can we resolve it? May need help from how webpack resolve it
- we need to make sure the object name from remote and host app are the same - I still not have an idea to do it

**Comment by flyfishzy** [comment]
@zgid123 , you are correct. Currently we need to ensure that host and remote have the same reference. This also seems reasonable for rollup, and maybe we need some help from the rollup community.

**Comment by zgid123** [comment]
or we can have a fallback? Like when the import raises error, we will use the file from the remote (I don't think it is possible), but I will give it a try. But recently I am too busy, I will try it later
